Add Long-form API documentation

jatkinson1000 opened the pull request:
Add long form API documentation using FORD

- [ ] Create GH action
- [ ] Creage GH page under ICCS
- [ ] Run FORD on code
- [ ] Update source with comments for FORD
- [ ] Move action to rebuild docs for main, not branch

jatkinson1000 commented:
Currently building at https://cambridge-iccs.github.io/FTorch/
Deployment seems to be working so now need to work on docs.

jatkinson1000 commented:
Suggestions for additional pages:

- [ ] Dome pleasant documentation for the examples
- [ ] Methods of installing libtorch/pytorch and where to find it onn the system etc - can be more verbose than README

It was also suggested that we reduce the README to make it more manageable whilst maintaining the key content there.

TomMelt commented:
According to https://forddocs.readthedocs.io/en/latest/user_guide/writing_documentation.html#non-fortran-source-files we should be able to use FORD for the C code. Not sure how well it works though

jatkinson1000 commented:
Indeed, I am looking at this locally and will push once I have something.

ElliottKasoar commented:
(Mostly a note for myself)

In the GPU section, it might be worth adding something about moving non- parameters/buffers to GPU, as I intend to do in Cambridge-ICCS/FTorch-benchmarks#27.

jatkinson1000 commented:
> (Mostly a note for myself)
> 
> In the GPU section, it might be worth adding something about moving non- parameters/buffers to GPU, as I intend to do in [Cambridge-ICCS/FTorch-benchmarks#27](https://github.com/Cambridge-ICCS/FTorch-benchmarks/pull/27).

I have added a GPU page - feel free to add to it - maybe on a branch if we want to be safe and then merge back into here.

jwallwork23 commented:
As discussed offline, perhaps we could work towards getting this PR merged and open new issues for the remaining checklist items? The content here is already a really good starting point. I'd be happy to review this PR if that'd be helpful?

jatkinson1000 commented:
> As discussed offline, perhaps we could work towards getting this PR merged and open new issues for the remaining checklist items? The content here is already a really good starting point. I'd be happy to review this PR if that'd be helpful?

Yes, please do!

jwallwork23 commented on pages/cmake.md:
```diff
@@ -0,0 +1,129 @@
+
+## Basic instructions
+
+To build the library, first clone it from github to your local machine and then run:
```
GitHub

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/cmake.md:
```diff
@@ -0,0 +1,129 @@
+
+Then invoke CMake with the Release build option, plus any other options as required
+from the table below in [CMake build options](#cmake-build-options)
+(note: you will likely _need_ to add some of these options to ):
```
Is this sentence missing something? "add some of these options to" what? Or, should the "to" be dropped?

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/cmake.md:
````diff
@@ -0,0 +1,129 @@
+cmake .. -DCMAKE_BUILD_TYPE=Release
+```
+
+Finally build and install the library:
+```bash
+cmake --build . --target install --config Release
+```
````
This differs slightly from what it says in the README:

> ```bash
> cmake --build . --target install
> ```
> or, if you want to separate these steps:
> ```bash
> cmake --build .
> cmake --install .
> ```

It would be good to be consistent.

jwallwork23 replied:
I wonder if we could even lift the installation instructions from a single source and insert it in both places?

jatkinson1000 replied:
Yes, that would be nice.
One thing we had discussed was that the README was getting rather unwieldly so as part of this PR (or a follow-up) we'd do some heavy pruning and point people here.

The install instructions would be one of the things that'd stay, however.

jatkinson1000 replied:
Agree on consistency. Just been trying to track down the discrepancy and I think it comes from when we were troubleshooting Windows builds. My understanding is that you can't use cache flags on windows so have to specify build type like that.

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/cmake.md:
```diff
@@ -0,0 +1,129 @@
+| `ENABLE_CUDA`                                                                                     | `TRUE` / `FALSE`             | Specifies whether to check for and enable CUDA<sup>2</sup> |
+
+
+<sup>1</sup> _The path to the Torch installation needs to allow cmake to locate the relevant Torch cmake files.  
```
CMake

jwallwork23 replied:
Also elsewhere in the file.

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/cmake.md:
````diff
@@ -0,0 +1,129 @@
+
+For example, to build on a unix system using the gnu compilers and install to `$HOME/FTorchbin/`
+we would need to run:
+```bash
+cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=icc -DCMAKE_CXX_COMPILER=icpc -DCMAKE_PREFIX_PATH=/path/to/venv/lib/python3.11/site-packages/torch/ -DCMAKE_INSTALL_PREFIX=~/FTorchbin
+```
````
This would be more readable as a multi-line statement using the `\` separator.

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/cmake.md:
````diff
@@ -0,0 +1,129 @@
+cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=icc -DCMAKE_CXX_COMPILER=icpc -DCMAKE_PREFIX_PATH=/path/to/venv/lib/python3.11/site-packages/torch/ -DCMAKE_INSTALL_PREFIX=~/FTorchbin
+```
+
+Once this completes you should be able to generate the code and install using:
+```bash
+cmake --build . --target install --config Release
+```
+or, if you want to separate these into two steps:
+```bash
+cmake --build .
+cmake --install . --config Release
+```
````
This is repeated. I guess we can drop the one from before this section?

jwallwork23 replied:
Again, note that the code here is slightly inconsistent with the README.

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/examples.md:
```diff
@@ -0,0 +1,13 @@
+
+## Worked examples
+
+#### 1) Simple
+
+#### 2) Resnet
+
+#### 3) 
```
Perhaps some placeholder text would be sufficient for now and we can open an issue for this section?

jatkinson1000 replied:
Examples updated in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/b9a633eb61dc5382b88ea715afb31de57551efab

jwallwork23 replied:
This looks great! My only minor comment is to use FTorch and CMake consistently.

jatkinson1000 replied:
Should now be done in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/a046727bebd77067cb48f02b6bb73619cb3659c8

jwallwork23 commented on pages/gpu.md:
This all reads well to me. Once merged, I'll add some docs on specifying device indices as part of #96.

jwallwork23 commented on pages/transposing.md:
```diff
@@ -0,0 +1,181 @@
+
+Transposition of data between Fortran and C can lead to a lot of unneccessary confusion.
+The FTorch library looks after this for you with the
+[`torch_tensor_from_array()`](doc/proc/torch_tensor_from_array.html) function which
```
This link gives me 404, is that expected?

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/b8ea751223de07f4f7c569bc3a811e960bfce20f

jwallwork23 commented on pages/transposing.md:
```diff
@@ -0,0 +1,181 @@
+
+This tells us how data output by Fortran should be read/accessed by Torch.
+
+TODO: Finish this
+passing `layout = [1, 2]` means that the data will be read in the correct indices by
+Torch.
+
+Tells us which order to read the indices in.
+i.e. `[1, 2]` will read `i` then `j`.
```
I guess this needs to be finished.

jatkinson1000 replied:
Think this is now done in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/b8ea751223de07f4f7c569bc3a811e960bfce20f but would appreciate a second pair of eyes reading as this always causes a headache, and worsens the more you think about it!

jwallwork23 commented on pages/transposing.md:
```diff
@@ -0,0 +1,181 @@
+you can take data from Fortran and send it to Torch to be indexed in exactly
+the same way.
+
+i.e. if the Fortran array `A`
+$$
+\begin{pmatrix}
+a & b \\
+c & d
+\end{pmatrix}
+$$
+is passed as `torch_tensor_from_array(A, )`
+the resulting Tensor will be read by Torch as 
```
I guess this also needs finishing.

jatkinson1000 replied:
As with above comment, should be done in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/b8ea751223de07f4f7c569bc3a811e960bfce20f but would appreciate a review/read.

jwallwork23 commented on pages/transposing.md:
```diff
@@ -0,0 +1,181 @@
+generate strides assuming this is what should happen.
+
+
+## Advanced use
+
+Those experienced with C will perhaps have noticed that there are further freedoms
+available beyond those presented above.
+
+[WIP]
+
+Always stride, or is there a transpose tradeoff?
```
If this is WIP then I recommend we open an issue for it and drop the text from here.

jatkinson1000 replied:
Opened #97 and removed the text here.

jwallwork23 commented on pages/troubleshooting.md:
```diff
@@ -0,0 +1,31 @@
+If possible we recommend using the [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/) (WSL) to build the library.
+In this case the build process is the same as for a Linux environment.
+
+If you need 
```
Can this just be dropped or is there something important to add?

jatkinson1000 replied:
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0

jwallwork23 commented on pages/troubleshooting.md:
```diff
@@ -0,0 +1,31 @@
+* [Visual Studio](https://visualstudio.microsoft.com/)
+* [Intel OneAPI Base and HPC toolkit](https://www.intel.com/content/www/us/en/developer/tools/oneapi/base-toolkit-download.html) (ensure that the Intel Fortran compiler and VS integration is selected).
+
+You should then be able to build from CMD following the 
```
Following what?

jatkinson1000 replied:
commands that I have just added in Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/90f1dfda90124cb93efa9acb6094c69c825184a0 😜

jwallwork23 reviewed (changes requested):
This is a really great start - lots of useful content, the FORD setup seems to work well, and the webpage looks nice. I checked all the links I came across and all worked except one (see comments).

Most of my requests are fairly minor and I'd be happy to help resolving them, if that would be useful?

I wonder if something like 'User guide' might be a better heading than 'Other documentation'? Just because there's lots of useful information in there.

jatkinson1000 commented on .github/workflows/build_docs.yml:
```diff
@@ -0,0 +1,56 @@
+          folder: doc      # The folder the action should deploy.
+          dry-run: true    # Don't actually push to pages, just test if we can
+
+      - name: Deploy Documentation for master
```
Should be fixed in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/37e15aee84e6ded3a05184c6385db6e6fedae673

jatkinson1000 commented:
Thanks @jwallwork23 I think those points are all covered in recent commits, including the moving 'Other docs' to 'User Guide'.

Also updated the landing page with latest resources and modified the workflow for `main` in anticipation of merging.

After casting an eye over I think this is probably now good to go subject to a re-review.

jatkinson1000 commented on .github/workflows/build_docs.yml:
```diff
@@ -0,0 +1,56 @@
+
+      - name: Deploy Documentation for master
+        # if: github.ref == 'refs/heads/main'
+        if: github.ref == 'refs/heads/API-docs'
```
Changed in anticipation of merging in https://github.com/Cambridge-ICCS/FTorch/pull/53/commits/821b301a11865ff9e1135c3f98880fbb2c239f23

jatkinson1000 commented:
This will also now close #98 as I have removed/simplified large chunks of the README to point at the online documentation.

jwallwork23 reviewed (approved):
Thanks very much for making these changes @jatkinson1000! The docs are now a great source of info for getting started, as well as covering some more advanced topics.

My only final request is very minor (see the unresolved thread above) so I'm happy to approve.
